### PR TITLE
planner, session: use SessionExtendedInfoSchema if snapshot isn't set

### LINKED ISSUE #38791: tidb panic when modifying column while doing some DML operations.
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
Use sysbench with oltp_writeonly workload, and modify the column type. 

### 2. What did you expect to see? (Required)
No error

### 3. What did you see instead (Required)
It may panic sometime.

### 4. What is your TiDB version? (Required)
6.3.0 master

### PR DESCRIPTION
**Pull request by wjhuang2016**
Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

### What problem does this PR solve?

Issue Number: close #38791

Problem Summary:
1. Previously, we `new` a SessionExtendedInfoSchema in tryLockMDLAndUpdateSchemaIfNecessary(). But it doesn't change the pass-in infoSchema. The compiler/Executor may use the InfoSchema instead of SessionExtendedInfoSchema later.
2. If the schema is changed, we need to clean TableToColumnMaps.

### What is changed and how it works?
1. When getting the session infoschema, `new` a SessionExtendedInfoSchema in GetDomainInfoSchema(). So that we don't need to new it tryLockMDLAndUpdateSchemaIfNecessary().
2. Clean TableToColumnMaps if the schema is changed.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
Sysbench with oltp_writeonly workload on a TiDB cluster, and `modify column` many times, no panic occurred.
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Fix a bug that connection panic when modifying a column while doing some DML operations with MDL enabled.
```

### CODE REVIEW

File: planner/core/preprocess.go [hunk 1]
```diff
@@ -1879,15 +1880,20 @@ func tryLockMDLAndUpdateSchemaIfNecessary(sctx sessionctx.Context, dbName model.
 
 		se, ok := is.(*infoschema.SessionExtendedInfoSchema)
 		if !ok {
-			se = infoschema.AttachMDLTableInfoSchema(is).(*infoschema.SessionExtendedInfoSchema)
-			sessiontxn.GetTxnManager(sctx).SetTxnInfoSchema(se)
-			sctx.GetSessionVars().TxnCtx.InfoSchema = se
+			panic("should not happen")
```
Comment (wjhuang2016) [inline planner/core/preprocess.go:1883]: test CI

Comment (lcwangchao) [reply planner/core/preprocess.go:1880]: why not just return  an error?

**Review by lcwangchao — APPROVED** [review]

**Review by tangenta — APPROVED** [review]

**Pull request merged**